### PR TITLE
7 model selection

### LINKED ISSUE #7: Model Selection
**Acceptance Criteria**:
- At least two real models are created (implementing `MirrorModel`), at least one of which requires a constructor parameter
- A `--model` required, named flag is added to the main script. This takes in the name of a model class. An instance of that model is created and passed to the fit method of the trainer.
  - If the model requires a constructor parameter, this would also be required as a CLI flag, like `--model.the_param`.
- The datasets read tokenized data from a subfolder of `~/nobackup/autodelete/mirror_data/tokenized`
- The model is initialized under the `trainer.fabric.init_module()` context
- `PlaceholderModel` is removed

### PR DESCRIPTION
**Pull request by JansenNye**
Ran test training runs using Wikitext dataset with LLaMa-3.2-1B (instruct and non-instruct) and GPT-2 implemented, including testing download/cache functionality and random weight initialization.
Closes #7

### CODE REVIEW

File: src/mirror/models/model_util.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        shutil.rmtree(model_path, ignore_errors=True)
+
+    path_exists = model_path.exists()
+    weights_present = path_exists and (any((model_path / name).exists() for name in (
+                "model.safetensors", "model.safetensors.index.json",
+                "pytorch_model.bin", "pytorch_model.bin.index.json",
+            )) 
+            or any(model_path.glob("model-*.safetensors")) 
+            or any(model_path.glob("pytorch_model-*.bin"))
+    )
```
Comment (Plyb) [inline src/mirror/models/model_util.py]: hmm, I don't love that we have to check this manually, it seems a bit brittle. Ideally there would be some huggingface-provided function we could use. If not though, this is fine.

File: src/mirror/tokenizers/mirror_gpt_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+class MirrorGPTTokenizer(MirrorTokenizer):
+    def __init__(self, hf_model_name: str) -> None:
+        self._hf_model_name = hf_model_name
+        self._tokenizer = load_hf_tokenizer_from_cache_or_download(hf_model_name)
+        if self._tokenizer.pad_token_id is None:
+            self._tokenizer.pad_token = self._tokenizer.eos_token
+
+    @property
+    def tokenization_id(self):
+        return self._hf_model_name
```
Comment (Plyb) [inline src/mirror/tokenizers/mirror_gpt_tokenizer.py]: Something to consider: the tokenization id is used to prevent duplicates when caching the tokenized dataset. If two models use the same tokenization scheme, this will cause them to have separate tokenized dataset caches, which would use more storage than needed. Is there a way we can prevent that?

File: src/mirror/models/mirror_llama_model.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        super().__init__()
+        hf_model_name = f"meta-llama/Llama-{id}"
+        if weights == "pretrained":
+            self.model = load_hf_model_from_cache_or_download(
+                hf_model_name,
+                model_cls=AutoModelForCausalLM,
+            )
+        elif weights == "random":
+            config = load_hf_config_from_cache_or_download(hf_model_name)
+            self.model = AutoModelForCausalLM.from_config(config)
```
Comment (Plyb) [inline src/mirror/models/mirror_llama_model.py]: This is complicated enough that that I wonder if it merits its own helper function so that it isn't duplicated between models. Borderline for me.

File: src/mirror/util.py [hunk 4]
```diff
@@ -44,3 +59,19 @@ def pad_to_longest(batch: list[TokenTensor], pad_token: int) -> tuple[TokenBatch
         tokens[i,:L] = b
     
     return tokens, attention_mask
+
+def instantiate_model(model: object, *, fabric):
```
Comment (Plyb) [inline src/mirror/util.py]: could give fabric a type, and the return type of the function here.

Comment (Plyb) [inline src/mirror/util.py]: `jsonargparse` provides an `instantiate_classes` function. Could we use that instead of making our own implementation?

File: src/main.py [hunk 5]
```diff
@@ -29,16 +30,20 @@
 import lightning.fabric.strategies
 import mirror.callbacks
 import mirror.datasets
+import mirror.models.mirror_gpt_model
+import mirror.models.mirror_llama_model
```
Comment (Plyb) [inline src/main.py]: Instead of doing this, create an `__init__.py` file under `models` and import that, like we're doing with `datasets` and `callbacks`.

File: src/main.py [hunk 6]
```diff
@@ -131,13 +142,30 @@ def _submit_slurm_job(*, python_args: list[str], slurm: SlurmConfig, num_nodes:
...
     if hasattr(cfg, 'config'):
         del cfg.config  # pyright: ignore
 
     init = parser.instantiate_classes(cfg)
-
-    main(**init)
+    if isinstance(init, dict):
+        main(**init)
+    else:
+        main(**vars(init))
```
Comment (Plyb) [inline src/main.py]: So, overall with this, here are my thoughts:
- the way our main script is currently set up (pre-this ticket) makes the model instantiation tricky
  - We need to defer instantiating the model until we have fabric set up so we can do it under `init_args`
  - But also we need to have model instantiated by the time we call main, or at least fabric
  - The main function is also set up in kind of a dumb way (my bad), because not all subcommands need all the parameters of all the other ones.
- My thoughts to fix it:
  - Change the `main` function so that it *only* takes in subcommand
    - then, within each case of the match statement, create a second parser for the parameters needed for *just that subcommand* (so the parameters of the `fit` function for the "fit" subcommand in this case)
  - Modify `fit`, so that instead of taking in `strategy`, `devices`, `num_nodes`, and `callbacks`, it takes in a `Trainer` instance.
  - Inside the the "fit" subcommand's case, do something a bit like what you're doing in the `__main__` script here:
    - add all of `fit`'s parameters *except* `model`
    - add the model separately so that we don't have to instantiate it when we do `instantiate_classes`.
    - instantiate everything except the model (note that this will include the trainer)
    - use the trainer to instantiate the model under `init_module`.
    - pass everything into `fit`

I'm considering whether or not we should do the first part (breaking up the parameter parsing into the case branches) in its own ticket, since that will be useful for #8, and we can get that out first. If that's too much of a pain though, we can just leave it in here.

File: src/mirror/models/mirror_llama_model.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    def preprocess_example(self, text: str) -> TokenTensor:
+        return self.tokenizer.encode(text)
+
+    def collate(self, examples: list[TokenTensor]) -> tuple[TokenBatch, AttentionMaskBatch]:
+        return pad_to_longest(examples, pad_token=self.tokenizer.pad_token_id)
+
+    def training_step(self, batch: tuple[TokenBatch, AttentionMaskBatch]) -> Loss:
+        input_ids, attention_mask = batch
+        labels = input_ids.clone()
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/mirror/models/mirror_llama_model.py]: We need to make labels separate from input_ids because we then modify labels to set padded positions to -100 and we don't want to set input_ids to -100 too

File: src/mirror/models/mirror_llama_model.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        if attention_mask is not None:
+            labels = labels.masked_fill(attention_mask == 0, -100)
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/mirror/models/mirror_llama_model.py]: attention_mask only tells the transformer what tokens to attend to - it doesn't stop the loss function from calculating loss at those positions, so if we didn't do this we'd calculate loss on padding tokens and basically be training the model to predict padding.
-100 is transformers' ignore_index; it's doing what the attention mask doesn't, telling the loss function not to calculate loss on these positions.

File: src/mirror/models/model_util.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+from mirror.util import assert_can_download, mirror_data_path
+
+
+def load_hf_model_from_cache_or_download(
+        hf_model_name: str | None = None,
+        reset_cache: bool = False,
+        model_cls: Type[PreTrainedModel] = AutoModel,
+) -> PreTrainedModel:
+    """
+    The first time this is called with a particular path/name pair, it will download
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/mirror/models/model_util.py]: Ah, yeah, forgot to change the docstring

File: src/main.py [hunk 10]
```diff
@@ -65,16 +70,23 @@ def main(
         job_id = _submit_slurm_job(python_args=sys.argv[1:], slurm=slurm, num_nodes=num_nodes, devices=devices)
         print(f"Submitted batch job {job_id}")
         return
+
+    if is_login_node() and not slurm.submit and subcommand == "fit":
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/main.py]: I initially had it implemented this way and it wasn't downloading properly, but I can't imagine why it wouldn't work so I'll try again

File: src/mirror/models/model_util.py [hunk 11]
```diff
@@ -14,61 +14,63 @@ def load_hf_model_from_cache_or_download(
...
-    config.save_pretrained(model_path)
-    return config
+    # Avoid creating config-only directory 
+    load_hf_model_from_cache_or_download(hf_model_name=hf_model_name, reset_cache=reset_cache)
+
+    return AutoConfig.from_pretrained(model_path, local_files_only=True)
+
+def build_causal_lm(
+        model_name: str | None = None,
+        weights: str = "pretrained"
```
Comment (Plyb) [inline src/mirror/models/model_util.py]: we can be a bit more specific here by saying `weights: Literal["pretrained", "random"]` (which is safer too, otherwise putting in a different string would result in returning `None`. Or maybe an exception, not sure)

File: src/mirror/models/model_util.py [hunk 12]
```diff
@@ -14,61 +14,63 @@ def load_hf_model_from_cache_or_download(
...
-    model_path.mkdir(parents=True, exist_ok=True)
-    config.save_pretrained(model_path)
-    return config
+    # Avoid creating config-only directory 
+    load_hf_model_from_cache_or_download(hf_model_name=hf_model_name, reset_cache=reset_cache)
+
+    return AutoConfig.from_pretrained(model_path, local_files_only=True)
+
+def build_causal_lm(
+        model_name: str | None = None,
```
Comment (Plyb) [inline src/mirror/models/model_util.py]: Why allow `None` here?

File: src/mirror/tokenizers/mirror_gpt_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+from mirror.tokenizers.tokenizer_util import load_hf_tokenizer_from_cache_or_download
+from mirror.util import get_device
+
+
+class MirrorGPTTokenizer(MirrorTokenizer):
+    def __init__(self, hf_model_name: str) -> None:
+        self._hf_model_name = hf_model_name
+        self._tokenizer = load_hf_tokenizer_from_cache_or_download(hf_model_name)
+        if self._tokenizer.pad_token_id is None:
+            self._tokenizer.pad_token = self._tokenizer.eos_token
```
[the remark this answers is not in the document]
Comment (Plyb) [reply src/mirror/tokenizers/mirror_gpt_tokenizer.py]: Ah, didn't realize it had both. But `pad_token_id` is the one being checked in the `if` condition, and the one that's actually being used below. I imagine we do want to set the `pad_token_id` at least at some point.

File: src/mirror/tokenizers/mirror_gpt_tokenizer.py [hunk 14]
```diff
@@ -1,34 +1,36 @@
...
-        return self._hf_model_name
+    def tokenization_id(self) -> str:
+        return "MirrorGPTTokenizer"
 
-    def encode(self, text):
+    def encode(self, text: str) -> TokenTensor:
         ids = self._tokenizer.encode(text, add_special_tokens=True)
         if len(ids) < 2:
             eos = self._tokenizer.eos_token_id
             eos = int(eos)
```
Comment (Plyb) [inline src/mirror/tokenizers/mirror_gpt_tokenizer.py]: is `eos_token_id` ever anything *but* an int?

File: src/mirror/tokenizers/tokenizer_util.py [hunk 15]
```diff
@@ -1,33 +1,32 @@
...
+tokenizers_path = mirror_data_path / "tokenizers"
 
 
-def load_hf_tokenizer_from_cache_or_download(
+def load_hf_tokenizer_from_cache_or_download[HfTokenizerT: PreTrainedTokenizerBase](
         hf_model_name: str | None = None,
         reset_cache: bool = False,
-        tokenizer_cls: Type[PreTrainedTokenizerBase] = AutoTokenizer,
-) -> PreTrainedTokenizerBase:
+) -> HfTokenizerT:
```
Comment (Plyb) [inline src/mirror/tokenizers/tokenizer_util.py]: python is weird with its generics. You can't explicitly pass a type argument to a function, the only way is to have it be inferred from the (value) parameters. So in this case, `HfTokenizerT` is always just going to be the most general type allowed (`PreTrainedTokenizerBase`). Might as well remove the type parameter.

**Review by Plyb — CHANGES_REQUESTED** [review]
A few comments from last time weren't addressed. One thing I usually do before sending something into review is to review the PR myself first. That tends to catch that kind of thing. I've resolved all the comments that were fixed. Overall great work! I'm liking it.

File: src/mirror/util.py (hunk 4 above)
Comment (JansenNye) [reply src/mirror/util.py]: Yes, it's just a little tricky because of initializing the model inside the fabric.trainer.init_module() context, but I'll implement that

File: src/main.py [hunk 16]
```diff
@@ -65,16 +70,23 @@ def main(
...
+        trainer = Trainer(strategy, devices, num_nodes, callbacks)
+        trainer.launch()
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/main.py]: Gotcha, I'll separate the login node/model download flow from the actual fitting one

File: src/mirror/models/model_util.py (hunk 12 above)
Comment (JansenNye) [reply src/mirror/models/model_util.py]: Yeah, that wouldn't ever be a valid input, I'll remove it

File: src/mirror/models/model_util.py [hunk 17]
```diff
@@ -1,16 +1,17 @@
...
+from jsonargparse import ArgumentParser, Namespace
 from lightning import Fabric
 from transformers import AutoConfig, AutoModel, AutoModelForCausalLM, PreTrainedModel, PretrainedConfig
-from typing import Type
+from typing import Literal, Type
 
 from mirror.download_util import assert_can_download, mirror_data_path
 from mirror.models.mirror_model import MirrorModel
 
+ignore_id = -100
```
Comment (Plyb) [inline src/mirror/models/model_util.py]: minor style thing, I think it's more common to have constants be UPPER_SNAKE_CASE

File: src/main.py [hunk 18]
```diff
@@ -56,16 +59,11 @@ def main(subcommand: Subcommand):
...
-                trainer=init.trainer,
-                checkpoint=init.checkpoint,
-                slurm=init.slurm,
-                epochs=init.epochs,
-                batch_size=init.batch_size,
-            )
+            init_cfg = cfg.clone()
+            del init_cfg.device
+            init = parser.instantiate_classes(init_cfg)
+            fit_kwargs = {name: getattr(init, name) for name in signature(fit).parameters}
```
Comment (Plyb) [inline src/main.py]: I think you only need either the `del` or the `signature` thing, not both, unless there's something I'm missing.

File: src/main.py [hunk 19]
```diff
@@ -57,40 +42,72 @@ def main(
...
+            run_config_yaml = f"subcommand: fit\n{parser.dump(cfg)}"
+
+            if hasattr(cfg, 'config'):
+                del cfg.config  # pyright: ignore
+
+            init_config(cfg.device)
+            init_cfg = cfg.clone()
+            del init_cfg.device
+            init = parser.instantiate_classes(init_cfg)
+            fit_kwargs = {name: getattr(init, name) for name in signature(fit).parameters}
```
Comment (Plyb) [inline src/main.py]: I think we only need either the `del` or the `signature` thing, not both.

File: src/main.py [hunk 20]
```diff
@@ -57,40 +42,72 @@ def main(
...
+    if is_login_node():
+        model = instantiate_model(model, fabric=None)
+        print("Model downloaded/cached. Re-run on a compute node.")
+        return
 
     trainer.launch()
 
-    with trainer.fabric.init_module():
-        model = PlaceholderModel()
+    model = instantiate_model(model, fabric=trainer.fabric)
```
Comment (Plyb) [inline src/main.py]: We should move the model instantiation stuff out of `fit`. As it stands, `fit` *says* it's accepting a `MirrorModel`, when it's actually accepting a namespace/dict that defines how to construct one.

Up in main, after we've done `instantate_classes`, `init.trainer.fabric` has a fabric instance which you can use to instantiate the model under `.init_module()`. After that is done, you can pass the fully initialized model in with all the other parameters to `fit`.

File: src/mirror/models/model_util.py [hunk 21]
```diff
@@ -79,15 +80,15 @@ def build_causal_lm(
...
+    model_parser.add_subclass_arguments(MirrorModel, "model", required=True, instantiate=True)
+
+    if fabric is not None:
+        def fabric_instantiator(class_type, *args, **kwargs):
+            with fabric.init_module():
+                return class_type(*args, **kwargs)
+
+        model_parser.add_instantiator(fabric_instantiator, MirrorModel, subclasses=True, prepend=True)
+
+    return model_parser.instantiate_classes(Namespace(model=model)).model
```
[the remark this answers is not in the document]
Comment (JansenNye) [reply src/mirror/models/model_util.py]: I thought we had to do the add_instantiator, but it seems you're right, I'll implement that
